external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -208,37 +412,37 @@ impl Compatibility {
         // friend list. But for simplicity, we decided to go to the more restrictive form now and
         // we may revisit this in the future.
         for (name, old_func) in &old_module.functions {
+            // if we don't find a function in the new module
             let Some(new_func) = new_module.functions.get(name) else {
                 if old_func.visibility == Visibility::Friend {
-                    friend_linking = false;
+                    context.function_missing_friend(old_func);
                 } else if old_func.visibility != Visibility::Private {
-                    datatype_and_function_linking = false;
+                    context.function_missing_public(old_func);
                 } else if old_func.is_entry && self.check_private_entry_linking {
-                    // This must be a private entry function. So set the link breakage if we're
-                    // checking for that.
-                    entry_linking = false;
+                    context.function_missing_entry(old_func);
                 }
                 continue;
             };
 
             // Check visibility compatibility
             match (old_func.visibility, new_func.visibility) {
                 (Visibility::Public, Visibility::Private | Visibility::Friend) => {
-                    datatype_and_function_linking = false
+                    context.function_lost_public_visibility(old_func);
+                }
+                (Visibility::Friend, Visibility::Private) => {
+                    context.function_lost_friend_visibility(old_func);
                 }
-                (Visibility::Friend, Visibility::Private) => friend_linking = false,
                 _ => (),
             }
 
             // Check entry compatibility
-            if old_module.file_format_version < VERSION_5
+            if (old_module.file_format_version < VERSION_5
                 && new_module.file_format_version < VERSION_5
-                && old_func.visibility != Visibility::Private
                 && old_func.is_entry != new_func.is_entry
+                && old_func.visibility != Visibility::Private)
+                || old_func.is_entry && !new_func.is_entry
```

Comment:
nit: It'd be good to keep these separate, so that we can clearly differentiate the old and new behaviour by file format version.

Reply:
+1, I'd like to keep this separate as well

Reply:
I'll have to add a clippy ignore here since it was identifying `if` and the `else if` as having the same effect.